tools/indexscan/main.go

```diff
@@ -251,6 +252,96 @@ func (s CommitsNearPublishStrategy) Search(ctx context.Context, r *git.Repositor
 	return
 }
 
+// ExhaustiveTreeSearchStrategy
+type ExhaustiveTreeSearchStrategy struct {
+}
+
+// Search returns the set of matching commits along with the number of matches.
+func (s ExhaustiveTreeSearchStrategy) Search(ctx context.Context, r *git.Repository, hashes []string) (closest []string, matched, total int, err error) {
+	// Find subset of file hashes that are present in the repo.
+	files := make(map[plumbing.Hash]int)
+	var i int
+	for _, h := range hashes {
+		_, err = r.BlobObject(plumbing.NewHash(h))
+		if err == plumbing.ErrObjectNotFound {
+			// Skip files not present in repo.
+			err = nil
+		} else if err != nil {
+			return
+		} else {
+			files[plumbing.NewHash(h)] = i
+			i++
+		}
+	}
+	// Iterate over repo trees to create an ordering.
+	trees := make(map[plumbing.Hash]int)
+	i = 0
+	for ti, _ := r.TreeObjects(); ; {
+		t, err := ti.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			panic(err)
+		} else {
+			trees[t.Hash] = i
+			i++
+		}
+	}
+	total = len(files)
+	if total == 0 {
+		err = errors.New("repo contains no matching files")
+		return
+	}
+	// Recursively build set of matching files for each tree.
+	matches := bitmap.NewBatch(len(files), len(trees))
+	processed := bitmap.New(len(trees))
+	ti, _ := r.TreeObjects()
+	ti.ForEach(func(t *object.Tree) error {
+		processTree(t, files, trees, matches, processed)
+		return nil
+	})
+	// Search through all commits for the one whose tree has the most matches.
+	ci, _ := r.CommitObjects()
+	err = ci.ForEach(func(c *object.Commit) error {
+		if !processed.Get(trees[c.TreeHash]) {
+			return errors.New("unprocessed tree")
+		}
+		count := matches[trees[c.TreeHash]].Count()
+		if matched < count {
+			matched = count
+			closest = closest[:0]
+		}
+		if matched == count {
+			closest = append(closest, c.Hash.String())
+		}
+		return nil
+	})
+	return
+}
+
+// processTree recursively determines the presence of a set of files in the given git Tree and records them.
+func processTree(t *object.Tree, files, trees map[plumbing.Hash]int, matches []bitmap.Bitmap, processed *bitmap.Bitmap) {
+	if processed.Get(trees[t.Hash]) {
+		return
+	}
+	for _, e := range t.Entries {
+		switch e.Mode {
+		case filemode.Dir:
+			if !processed.Get(trees[t.Hash]) {
```

Comment:
Isn't this conditional always true? There's a guard clause at the start of processTree that IIUC is making sure this is true.

Should this be `e.Hash` instead of `t.Hash`? Or maybe just delete and let the recursive call into `processTree` do that check?

Reply:
eh i'd prefer to avoid the call to Tree.Tree() if possible so I'll just fix the bug.